### ISSUE TITLE
Possible Unbound False Positive for Tuple Iteration

### ISSUE THREAD

**Issue by max-muoto** [microsoft/pyright#7299]
Let's say you had the following code:

```python
from typing import reveal_type


def foo(t: list[int]):
    for item in t:
        v = item
    reveal_type(v) # MyPy: builtins.int. Pyright: Possibly unbound.

foo([])
```

Here, MyPy is giving us a false negative, even though this could be unbound, given an empty list. However, if we have:

```python
from typing import reveal_type


def foo(t: tuple[int, ...]):
    for item in t:
        v = item
    reveal_type(v) # Pyright: Possibly unbound.

foo((2,))
```

[Playground Link
](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMogCmAboQIYA2A%2BvAoQFCMAmhwUwYYAFDAFywBXBBUIBtVDAA0UAHRyAugEpe9KGva5MMQtFSwV6w1GJQAvFp2r1RUpRqJCXYoqiMO3LgCZJixUA)

Shouldn't we be able to say `v` is an `int`, since we know there is at least one element in the tuple?

**Comment by max-muoto** [comment]
Edit: I'm seeing something that would further indicate why. Once again, there is a different between MyPy and Pyright here.

```python
tup: tuple[int] = tuple()
```

MyPy gives: Incompatible types in assignment (expression has type "tuple[Never, ...]", variable has type "tuple[int]") 

Pyright is fine with this.

**Comment by erictraut** [comment]
Pyright is working as intended in all of the above code samples. I don't see any bugs here.

The last example (`tup: tuple[int] = tuple()`) does not generate an error because of the way that the tuple constructor is defined in typeshed. If you're interested in making your code as robust as possible, I recommend avoiding calls to the tuple constructor and instead using tuple expressions. Type checkers like pyright (and mypy, for that matter) are very smart about tuple expressions.

The documentation you linked to is about tuple expression type inference. It isn't talking about tuple constructors or general use cases of tuples.
